### DISCUSSION TITLE
Deprecate iterators returned by `mdns::Event`

### DISCUSSION THREAD

**Discussion by drHuangMHT** [libp2p/rust-libp2p#3612] [Ideas]
Let's push #3593 and #3606 further. Since it feels wrong to clone single-use iterators, and you have to manually convert a `SmallVec` into an iterator before putting it into an event, why not get our hands on deprecating them? However, I cannot come up with a way to gently roll out the change and I guess it's not possible to do so. But adapting to the change for established code is relatively easy, you'll only need to add `iter()` or `into_iter()` to make things work again temporarily and re-design later.

**Comment by thomaseizinger** [comment]
I think this would just have to be a breaking change. Can you open a draft PR with the API you are thinking of?
